File: dotnet/src/Examples/BasicPromptingClean.cs
using AIPairProgramming.Shared.Models;

namespace AIPairProgramming.Examples;

public class BasicPromptingClean
{
    public async Task<ValidationResult<AuthToken>> AuthenticateUserAsync(LoginCredentials credentials)
    {
        throw new NotImplementedException();
    }

    public decimal CalculateDiscount(string loyaltyTier, decimal orderTotal)
    {
        throw new NotImplementedException();
    }

    public decimal GetDiscount(string tier, decimal total)
    {
        throw new NotImplementedException();
    }

    public class DiscountCalculator(IPricingStrategy pricingStrategy)
    {
        public async Task<ValidationResult<decimal>> CalculateCustomerDiscountAsync(string customerId, decimal orderAmount)
        {
            throw new NotImplementedException();
        }
    }

    public async Task<ValidationResult<User>> RegisterUserAsync(UserRegistration userData)
    {
        throw new NotImplementedException();
    }

    public async Task<ValidationResult<ShippingQuote>> GenerateShippingQuoteAsync(ShippingRequest request)
    {
        throw new NotImplementedException();
    }

    public decimal CalculateShipping(decimal weight, decimal distance)
    {
        throw new NotImplementedException();
    }

    public int CalculateShippingCost(decimal weight, decimal distance)
    {
        throw new NotImplementedException();
    }

    public async Task<ValidationResult<string>> ProcessPaymentAsync(decimal amount, string cardToken)
    {
        throw new NotImplementedException();
    }

    public ValidationResult<string> ValidateEmail(string email)
    {
        throw new NotImplementedException();
    }

    public class PaymentService(IPaymentGateway gateway, ILogger<PaymentService> logger)
    {
        public async Task<ValidationResult<string>> ProcessAsync(decimal amount)
        {
            throw new NotImplementedException();
        }
    }

    public record PaymentRequest(decimal Amount, string Currency, string CustomerId)
    {
        public string? Description { get; init; }
    }

    public decimal CalculateFees(PaymentRequest request) => request switch
    {
        { Amount: < 5 } => throw new ArgumentException("Minimum amount is $5"),
        { Amount: > 10000 } => CalculateLargeTransactionFee(request),
        { Currency: not "USD" } => CalculateInternationalFee(request),
        _ => CalculateStandardFee(request)
    };

    private decimal CalculateLargeTransactionFee(PaymentRequest request) => 0m;
    private decimal CalculateInternationalFee(PaymentRequest request) => 0m;
    private decimal CalculateStandardFee(PaymentRequest request) => 0m;

    public ValidationResult<string> ValidateEmailAI(string email)
    {
        if (string.IsNullOrEmpty(email))
        {
            return ValidationResult<string>.FailureResult("Email is required");
        }

        var emailRegex = new System.Text.RegularExpressions.Regex(@"^[^\s@]+@[^\s@]+\.[^\s@]+$");
        if (!emailRegex.IsMatch(email))
        {
            return ValidationResult<string>.FailureResult("Invalid email format");
        }

        return ValidationResult<string>.SuccessResult(email.ToLower().Trim());
    }

    public ValidationResult<string> ValidateEmailImproved(string email)
    {
        if (email is null)
        {
            return ValidationResult<string>.FailureResult("Email must not be null");
        }

        var trimmed = email.Trim();

        if (trimmed.Length == 0)
        {
            return ValidationResult<string>.FailureResult("Email is required");
        }

        if (trimmed.Length > 254)
        {
            return ValidationResult<string>.FailureResult("Email too long (max 254 characters)");
        }

        var emailRegex = new System.Text.RegularExpressions.Regex(
            @"^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$");

        if (!emailRegex.IsMatch(trimmed))
        {
            return ValidationResult<string>.FailureResult("Invalid email format");
        }

        var localPart = trimmed.Split('@')[0];
        if (localPart.Length > 64)
        {
            return ValidationResult<string>.FailureResult("Email local part too long (max 64 characters)");
        }

        return ValidationResult<string>.SuccessResult(trimmed.ToLower());
    }

    public void CheckForWrongApiUsage()
    {
    }

    public bool CheckBusinessLogic(int age)
    {
        return age >= 13;
    }

    public string[] CheckPerformance(string[] items)
    {
        var result = new List<string>(items.Length);
        foreach (var item in items)
        {
            if (!string.IsNullOrEmpty(item))
            {
                result.Add(item.ToUpper());
            }
        }
        return result.ToArray();
    }

    public ValidationResult<decimal> SafeDivision(decimal dividend, decimal divisor)
    {
        if (divisor == 0)
        {
            return ValidationResult<decimal>.FailureResult("Cannot divide by zero");
        }

        return ValidationResult<decimal>.SuccessResult(dividend / divisor);
    }

    public async Task<ValidationResult<PaymentConfirmation?>> ProcessPaymentAsync(
        PaymentRequest request,
        CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }

    private static readonly string[] SupportedCurrencies = ["USD", "EUR", "GBP"];

    public record PaymentConfirmation(string Id, decimal Amount, string Status);
    public record ShippingRequest(decimal Weight, decimal Distance, string Priority);
    public record ShippingQuote(decimal Cost, int EstimatedDays, string Carrier);

    public class PromptIterationExample
    {
        public bool ValidateInputV1(string input)
        {
            return !string.IsNullOrEmpty(input);
        }

        public ValidationResult<string> ValidateInputV2(string input)
        {
            if (string.IsNullOrEmpty(input))
                return ValidationResult<string>.FailureResult("Input required");
            
            return ValidationResult<string>.SuccessResult(input);
        }

        public ValidationResult<UserRegistration> ValidateRegistrationInput(string name, string email, string password)
        {
            var errors = new List<ValidationError>();

            if (string.IsNullOrWhiteSpace(name))
                errors.Add(new ValidationError(nameof(name), "REQUIRED", "Name is required"));
            else if (name.Length < 2 || name.Length > 100)
                errors.Add(new ValidationError(nameof(name), "LENGTH", "Name must be 2-100 characters"));
            else if (!System.Text.RegularExpressions.Regex.IsMatch(name, @"^[a-zA-Z\s]+$"))
                errors.Add(new ValidationError(nameof(name), "FORMAT", "Name can only contain letters and spaces"));

            var passwordErrors = ValidatePasswordStrength(password);
            errors.AddRange(passwordErrors);

            if (errors.Count > 0)
                return ValidationResult<UserRegistration>.FailureResult(errors);

            return ValidationResult<UserRegistration>.SuccessResult(new UserRegistration(email, name, password));
        }

        private List<ValidationError> ValidatePasswordStrength(string password)
        {
            var errors = new List<ValidationError>();
            
            if (string.IsNullOrEmpty(password))
                errors.Add(new ValidationError("password", "REQUIRED", "Password is required"));
            else
            {
                if (password.Length < 12)
                    errors.Add(new ValidationError("password", "TOO_SHORT", "Password must be at least 12 characters"));
                
                if (!password.Any(char.IsUpper))
                    errors.Add(new ValidationError("password", "NO_UPPERCASE", "Password must contain uppercase letter"));
                
                if (!password.Any(char.IsLower))
                    errors.Add(new ValidationError("password", "NO_LOWERCASE", "Password must contain lowercase letter"));
                
                if (!password.Any(char.IsDigit))
                    errors.Add(new ValidationError("password", "NO_DIGIT", "Password must contain number"));
                
                if (!password.Any(ch => !char.IsLetterOrDigit(ch)))
                    errors.Add(new ValidationError("password", "NO_SYMBOL", "Password must contain special character"));
            }
            
            return errors;
        }
    }
}

public interface IPricingStrategy
{
    Task<decimal> CalculateDiscountAsync(string customerId, decimal orderAmount);
}

public interface IPaymentGateway
{
    Task<ValidationResult<string>> ProcessPaymentAsync(decimal amount, string cardToken);
}

public interface ILogger<T>
{
}

File: nodejs/src/examples/basic-prompting-clean.ts
import { ValidationResult, LoginCredentials, AuthToken } from '../types/auth';
import { PricingStrategy } from '../types/pricing';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';


/**
 * Authenticates user with email and password
 *
 * Constraints:
 * - Email must be valid format (RFC 5322)
 * - Password must meet security requirements
 * - Rate limiting: max 5 attempts per 15 minutes per IP
 * - Uses bcrypt for password verification
 * - Returns JWT token valid for 24 hours
 *
 * Examples:
 * - Valid: {email: "user@example.com", password: "SecurePass123!"} → {success: true, token: "eyJ0..."}
 * - Invalid email: {email: "invalid-email", password: "password"} → {success: false, error: "Invalid email format"}
 * - Wrong password: {email: "user@example.com", password: "wrong"} → {success: false, error: "Invalid credentials"}
 */


function calculateDiscount(loyaltyTier: string, orderTotal: number): number {
  throw new Error('Not implemented');
}

function getDiscount(tier: string, total: number): number {
  throw new Error('Not implemented');
}

class DiscountCalculator {
  constructor(private pricingStrategy: PricingStrategy) {}

  calculateCustomerDiscount(customerId: string, orderAmount: number): Promise<ValidationResult<number>> {
    throw new Error('Not implemented');
  }
}

async function registerUser(userData: unknown): Promise<ValidationResult> {
  throw new Error('Not implemented');
}

function generateShippingQuote(request: unknown): Promise<ValidationResult> {
  throw new Error('Not implemented');
}

function calculateShipping(weight: number, distance: number): number {
  throw new Error('Not implemented');
}

function calculateShippingCost(weight: number, distance: number): number {
  throw new Error('Not implemented');
}

async function processPayment(amount: number, cardToken: string): Promise<ValidationResult> {
  throw new Error('Not implemented');
}

function validateEmail(email: string): ValidationResult<string> {
  throw new Error('Not implemented');
}

function validateEmailAI(email: string): ValidationResult<string> {
  if (!email) {
    return { success: false, error: 'Email is required' };
  }

  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  if (!emailRegex.test(email)) {
    return { success: false, error: 'Invalid email format' };
  }

  return {
    success: true,
    data: email.toLowerCase().trim()
  };
}

function validateEmailImproved(email: string): ValidationResult<string> {
  if (typeof email !== 'string') {
    return { success: false, error: 'Email must be a string' };
  }

  const trimmed = email.trim();

  if (trimmed.length === 0) {
    return { success: false, error: 'Email is required' };
  }

  if (trimmed.length > 254) {
    return { success: false, error: 'Email too long (max 254 characters)' };
  }

  const emailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;

  if (!emailRegex.test(trimmed)) {
    return { success: false, error: 'Invalid email format' };
  }

  const [localPart] = trimmed.split('@');
  if (localPart.length > 64) {
    return { success: false, error: 'Email local part too long (max 64 characters)' };
  }

  return {
    success: true,
    data: trimmed.toLowerCase()
  };
}

function checkForWrongApiUsage() {
  // Implementation
}

function checkBusinessLogic(age: number): boolean {
  return age >= 13;
}

function checkPerformance(items: string[]): string[] {
  const result: string[] = [];
  for (const item of items) {
    if (item.length > 0) {
      result.push(item.toUpperCase());
    }
  }
  return result;
}

// In-memory user store for demonstration
const users = [
  {
    email: 'user@example.com',
    // bcrypt hash for 'SecurePass123!'
    passwordHash: '$2a$10$wq8Qw1Q6l6Qw1Q6l6Qw1QeQw1Q6l6Qw1Q6l6Qw1Q6l6Qw1Q6l6Qw1Q',
    id: '1'
  }
];

// In-memory rate limit store: { [ip]: { count, firstAttempt: Date } }
const rateLimitStore: Record<string, { count: number; firstAttempt: number }> = {};

const PASSWORD_MIN_LENGTH = 8;
const PASSWORD_REGEX = /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d!@#$%^&*()_+\-=\[\]{};':"\\|,.<>\/?]{8,}$/;
const JWT_SECRET = 'your_jwt_secret'; // Replace with env var in production

/**
 * Authenticates user with email and password
 *
 * Constraints:
 * - Email must be valid format (RFC 5322)
 * - Password must meet security requirements
 * - Rate limiting: max 5 attempts per 15 minutes per IP
 * - Uses bcrypt for password verification
 * - Returns JWT token valid for 24 hours
 */
async function authenticateUser(
  credentials: LoginCredentials,
  ip: string
): Promise<ValidationResult<AuthToken>> {
  // Rate limiting
  const now = Date.now();
  const windowMs = 15 * 60 * 1000;
  const maxAttempts = 5;
  const rl = rateLimitStore[ip];
  if (rl) {
    if (now - rl.firstAttempt < windowMs) {
      if (rl.count >= maxAttempts) {
        return { success: false, error: 'Too many attempts. Try again later.' };
      }
      rl.count += 1;
    } else {
      rl.count = 1;
      rl.firstAttempt = now;
    }
  } else {
    rateLimitStore[ip] = { count: 1, firstAttempt: now };
  }

  // Email validation (RFC 5322)
  const emailResult = validateEmailImproved(credentials.email);
  if (!emailResult.success) {
    return { success: false, error: 'Invalid email format' };
  }

  // Password validation
  if (
    typeof credentials.password !== 'string' ||
    credentials.password.length < PASSWORD_MIN_LENGTH ||
    !PASSWORD_REGEX.test(credentials.password)
  ) {
    return { success: false, error: 'Password does not meet security requirements' };
  }

  // User lookup
  const user = users.find(
    (u) => u.email.toLowerCase() === credentials.email.toLowerCase()
  );
  if (!user) {
    return { success: false, error: 'Invalid credentials' };
  }

  // Password verification
  const match = await bcrypt.compare(credentials.password, user.passwordHash);
  if (!match) {
    return { success: false, error: 'Invalid credentials' };
  }

  // JWT generation
  const token = jwt.sign(
    { userId: user.id, email: user.email },
    JWT_SECRET,
    { expiresIn: '24h' }
  );

  return { success: true, data: { token } };
}

export {
  authenticateUser,
  calculateDiscount,
  getDiscount,
  DiscountCalculator,
  registerUser,
  generateShippingQuote,
  calculateShipping,
  calculateShippingCost,
  processPayment,
  validateEmail,
  validateEmailAI,
  validateEmailImproved,
  checkForWrongApiUsage,
  checkBusinessLogic,
  checkPerformance
};
